internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -180,3 +183,166 @@ func (*topic) IsRetryable(error) bool {
+// queue must have been previously created (for instance, by using
+// amqp.Channel.QueueDeclare) and bound to an exchange.
+//
+// OpenSubscription uses the supplied amqp.Connection for all communication. It is
+// the caller's responsibility to establish this connection before calling OpenTopic,
+// and to close it when Close has been called on all Subscriptions opened with it.
+//
+// The documentation of the amqp package recommends using separate connections for
+// publishing and subscribing.
+func OpenSubscription(conn *amqp.Connection, name string) (*pubsub.Subscription, error) {
+	// TODO(jba): support context.Context
+	s, err := newSubscription(conn, name)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewSubscription(s), nil
+}
+
+type subscription struct {
+	conn     *amqp.Connection
+	queue    string // the AMQP queue name
+	consumer string // the client-generated name for this particular subscriber
+
+	mu   sync.Mutex
+	ch   *amqp.Channel // AMQP channel used for all communication.
+	delc <-chan amqp.Delivery
+}
+
+var nextConsumer int64 // atomic
+
+func newSubscription(conn *amqp.Connection, name string) (*subscription, error) {
+	s := &subscription{
+		conn:     conn,
+		queue:    name,
+		consumer: fmt.Sprintf("c%d", atomic.AddInt64(&nextConsumer, 1)),
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if err := s.establishChannel(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// Must be called with s.mu held.
+func (s *subscription) establishChannel() error {
+	// TODO(jba): support context.Context
+	if s.ch != nil {
+		// We already have a channel; nothing to do.
+		return nil
+	}
+	// Create a new channel.
+	ch, err := s.conn.Channel()
+	if err != nil {
+		return err
+	}
+	// Subscribe to messages from the queue.
+	s.delc, err = ch.Consume(s.queue, s.consumer,
+		false, // autoAck
+		false, // exclusive
+		false, // noLocal
+		wait,
+		nil) // args
+	if err != nil {
+		return err
+	}
+	s.ch = ch
+	return nil
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	// TODO(jba): Improve the timing characteristics of this implementation:
+	// If messages come at a rate just below the timeout, then we can take a long time to gather maxMessages.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if err := s.establishChannel(); err != nil {
+		return nil, err
+	}
+
+	// Get all the waiting messages, up to maxMessages.
+	var ms []*driver.Message
+loop:
+	for len(ms) < maxMessages {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+
+		case d, ok := <-s.delc:
+			if !ok { // channel closed
+				s.ch = nil // re-establish the channel
+				if len(ms) > 0 {
+					return ms, nil
+				}
+				// TODO(jba): get more information for a better message
+				return nil, errors.New("delivery channel closed")
+			}
+			ms = append(ms, toMessage(d))
+		default:
```

Comment:
Why have the `default`? IIUC, you could leave out the `default` and `loop`, and the `select` would block on `delc`; isn't that what you want? To avoid blocking forever you could move your  `<-time.After()` case up here and get rid of the lower `select`.

Reply:
Yes, that's how I had it originally, and that's why I added the TODO at the top of the function (which I forgot to remove). 

In the old way, if a message comes in every 40ms and `maxMessages` is 100, ReceiveBatch will stay around for 4s, collecting 100 messages. With the current code, it will return one message immediately. I think that's better.

The story is probably not over, though. The channel returned by Consume (`delc`) is unbuffered, which may mean that we'll never get more than one message even if others are waiting somewhere. I've changed my TODO to reflect that.

Reply:
WDYT about creating a `time.After` channel outside the loop instead? Something like:

```
maxTime := time.After(50*time.Millisecond)
done := false
for !done && len(ms) < maxMessages {
  select {
    case <-ctx.Done(): ...
    case d, ok := <-s.delc: ...
    case <-maxTime: ... done = true; break
  }
}
return ms, nil
```

That way the total time of the call is bounded, but we'll be more likely to collect multiple messages that are delivered close together in time.

Reply:
WDYT?

Reply:
Messages might arrive every 26ms, but ReceiveBatch would return one message every 50ms, halving possible throughput.

Reply:
The first call to `ReceiveBatch` would take 50ms and return 1 message, but the 2nd call would take 50ms and return 2 messages, no?

Reply:
Ah, right. So maybe that's okay.

Made the change. PTAL.